doc/extending/using_params.txt

```diff
@@ -0,0 +1,140 @@
+Defining a params type
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. note::
+
+    This section is only relevant if you decide to create your own type.
+
+The first thing you need to do is to define a Theano Type for your
+params object.  It doesn't have to be complete type because only the
+following methods will be used for the type:
+
+  - :meth:`filter <PureType.filter>`
+  - :meth:`__eq__ <PureType.__eq__>`
+  - :meth:`__hash__ <PureType.__hash__>`
+  - :meth:`values_eq <PureType.values_eq>`
+
+Additionaly if you want to use your params with C code, you need the
+following methods:
+
+  - :meth:`c_declare <CLinkerType.c_declare>`
+  - :meth:`c_init <CLinkerType.c_init>`
+  - :meth:`c_extract <CLinkerType.c_extract>`
+  - :meth:`c_cleanup <CLinkerType.c_cleanup>`
+
+You can also define other convinience methods such as
+:meth:`c_headers <CLinkerType.c_headers>` if you need any special things.
+
+
+Registering the params with your Op
+-----------------------------------
+
+To declare that your op uses params you have to set the class
+attribute :attr:`params_type` to an instance of your params Type.
+
+For example if we decide to use a pointer to an int as the params the
+following would be appropriate:
+
+.. code-block:: python
+
+   class MyOp(Op):
+       params_type = CDataType('int')
+
+After that you need to define a :meth:`get_params` method on your
+class with the following signature:
+
+.. code-block:: python
+
+   def get_params(self, node):
+       pass
+
+This methods must return a valid object for your Type (an object that
+passes :meth:`filter`).  The `node` parameter is the Apply node for
+which we want the params.  Therefore the params object can depend on
+the inputs and outputs of the node.
+
+.. note::
+
+    Due to implementation restrictions, None is not allowed as a
+    params object and will be taken to mean that the Op doesn't have
+    parameters.
+
+    Since this will change the expected signature of a few methods, it
+    is strongly discouraged to have your :meth:`get_params` method
+    return None.
+
+
+Signature changes from having params
+------------------------------------
+
+Having declared a params for you Op will affect the expected signature
+of :meth:`perform`.  The new expected signature will have an extra
+parameter at the end which corresponds to the params object.
+
+.. warning::
+
+   If you do not account for this extra parameter, the code will fail
+   at runtime if it tries to run the python version.
+
+Also, for the C code, the `sub` dictionary will contain an extra entry
+`'params'` which will map to the variable name of the params object.
+This is true for all methods that recieve a `sub` parameter, so this
+means that you can use your params in the :meth:`c_init_code_struct
+<Op.c_init_code_struct>` method.
+
+A simple example
+----------------
+
+This is a simple example which uses a params object to pass a value.
+The value in this case is a python float, hence the choice of Generic
+as the params type.
```

Comment:
State the functionality implemented by this example, either here or in a class comment.